src/aspects/sagemaker-automated-shutdown/index.ts

```diff
@@ -0,0 +1,99 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { Stack, IAspect } from 'aws-cdk-lib';
+import {
+    CfnNotebookInstance,
+    CfnNotebookInstanceLifecycleConfig
+} from 'aws-cdk-lib/aws-sagemaker';
+import { IConstruct } from 'constructs';
+import { readFileSync } from 'fs';
+import { join } from 'path';
+
+export interface SageMakerShutdownProps {
+    readonly idleTimeoutMinutes?: number;
+}
+
+export class SageMakerAutomatedShutdown implements IAspect {
+    private processedInstances: Set<string> = new Set();
+
+    constructor(private readonly props: SageMakerShutdownProps = {}) {}
+
+    visit(node: IConstruct): void {
+        if (!(node instanceof CfnNotebookInstance)) {
+            return;
+        }
+
+        const cfnInstance = node as CfnNotebookInstance;
+        if (this.processedInstances.has(cfnInstance.ref)) {
+            return;
+        }
+
+        const stack = Stack.of(node);
+        this.applyShutdownMechanism(cfnInstance, stack);
+        this.processedInstances.add(cfnInstance.ref);
+    }
+
+    private applyShutdownMechanism(
+        instance: CfnNotebookInstance,
+        stack: Stack
+    ): void {
+        const lifecycleConfig = this.createLifecycleConfig(instance, stack);
+        instance.lifecycleConfigName = lifecycleConfig.ref;
```

Comment:
This will overwrite the existing lifecycle config if there is one right? Is there a way to append to an existing lifecycle script if the user has already defined one? I didn't try this when we used this before but if we can't, we need to define a clear docstring that tells the user that this will overwrite any existing lifecycle configuration policy they've defined